From 424ead2746639916ad6ded411d08c3d18f109520 Mon Sep 17 00:00:00 2001
From: Franck LENORMAND <franck.lenormand@nxp.com>
Date: Wed, 4 Dec 2019 12:22:24 +0100
Subject: [PATCH 62/87] TEE-545-1 Instantiate RNG state handle with Prediction
 Resistance support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Instantiate RNG state handles with Prediction Resistance (PR) support.
This way SW further downstream (e.g. Rich OS, boot loader etc.) is able
to use the "PR" bit in RNG generation descriptors (forcing TRNG
re-seeding before PRNG / DRBG outputs random data).

Note: current patch does not deal with RNG state handles that have
already been initialized, but without PR support (this could happen if
U-boot would run before OP-TEE etc.).
In this case, RNG state handle would have to be deinstantiated first,
and then reinstantiated with PR support.

Signed-off-by: Franck LENORMAND <franck.lenormand@nxp.com>
Signed-off-by: Horia GeantÄƒ <horia.geanta@nxp.com>
---
 core/drivers/crypto/caam/include/caam_desc_defines.h | 3 +++
 core/drivers/crypto/caam/include/caam_desc_helper.h  | 2 +-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/core/drivers/crypto/caam/include/caam_desc_defines.h b/core/drivers/crypto/caam/include/caam_desc_defines.h
index 2972a6a2..ac150e6b 100644
--- a/core/drivers/crypto/caam/include/caam_desc_defines.h
+++ b/core/drivers/crypto/caam/include/caam_desc_defines.h
@@ -417,6 +417,9 @@
 /* State Handle */
 #define ALGO_RNG_SH(sh)		SHIFT_U32((sh) & 0x3, 4)
 
+/* Prediction Resistance */
+#define ALGO_RNG_PR BIT32(1)
+
 /* State */
 #define AS_RNG_GENERATE		0x0
 #define AS_RNG_INSTANTIATE	0x1
diff --git a/core/drivers/crypto/caam/include/caam_desc_helper.h b/core/drivers/crypto/caam/include/caam_desc_helper.h
index dd9b5b3c..4f481ac9 100644
--- a/core/drivers/crypto/caam/include/caam_desc_helper.h
+++ b/core/drivers/crypto/caam/include/caam_desc_helper.h
@@ -264,7 +264,7 @@ static inline void dump_desc(uint32_t *desc)
  */
 #define RNG_SH_INST(sh)                                                        \
 	(CMD_OP_TYPE | OP_TYPE(CLASS1) | OP_ALGO(RNG) | ALGO_RNG_SH(sh) |      \
-	 ALGO_AS(RNG_INSTANTIATE))
+	 ALGO_AS(RNG_INSTANTIATE) | ALGO_RNG_PR)
 
 /*
  * RNG Generates Secure Keys
-- 
2.17.1

